frontend/src/app/forms/forms.page.ts

```diff
@@ -205,15 +212,23 @@ export class FormsPage implements OnInit, OnDestroy {
    * "Submit" the form: Store the value in our persistent storage and navigate to the confirmation page.
    */
   submit(): void {
-    const key = (
-      this.formsData.value.basicInfo.firstName +
-      '-' +
-      this.formsData.value.basicInfo.lastName +
-      '-' +
-      Math.floor(Math.random() * 1000000).toString()
-    ).replace(/\s+/g, '_');
-    this.storage.addItem(key, this.formsData.value);
-    this.router.navigate(['/forms', 'confirmation']);
+    const time = new Date().toLocaleTimeString('en-US', {
+      hour12: false,
+      hour: 'numeric',
+      minute: 'numeric',
+    });
+    this.formsData.controls.submittedTime.setValue(time);
+
+    this.storage.getNextId().then((key) => {
+      this.formsData.controls.id.setValue(key);
+
+      this.storage.addItem<FormsData>(
+        this.storage.applicantDetailsDb,
+        key,
+        this.formsData.getRawValue()
```

Comment:
In 99.9% of the cases, you don't actually want to use "getRawValue". Why is this necessary here? Please keep in mind that getRawValue will include all fields, including disabled ones. This quickly leads to bugs. Or it "leaks" data the user didn't even want to submit. So, one should be really really careful when using this.

Reply:
Ya I know what it is being used for and aware of it. The main reason is I have added id and isRated flag in the interface , I don't want these to validated as it is system generated. SO I disabled these fileds. But I want these fields in storage. Hence used getRawValue.

Reply:
> I don't want these to validated as it is system generated

If you don't add a Validator, then they won't be validated. ;) So I see no problem here...

The only potential issue I see at this point is that the progress bar would also "react" to these values. But we should rather fix the progress bar instead of doing sth like this.

Is there any other reason why this is necessary? If not, then I strongly recommend against using `getRawValue()`.

Reply:
Currently the interface is `export interface FormsData {
  id: string;
  isRated: number;
  submittedTime: string;
  basicInfo: BasicInfo;
  contactInfo: ContactInfo;
  educationInfo: EducationInfo;
  fieldDesignationInfo: FieldDesignationInfo;
}`

Id and isRated are not under FormGroup as like other fields. And I don't think it is correct to move as a Group. If you leave without Validators.required valid property becomes true. We are using value property of the first level Control object to calculate progress. If we did not disable it I have to make poor hack in the code to work. So I used disable() to remove it from control as suggested in https://angular.io/api/forms/AbstractControl#disable . But I need the disabed values in storage. hence I used getRawValues(). Also that is one of the usecase why  such methods exists.

Reply:
> We are using value property of the first level Control object to calculate progress.

Then we should definitely refactor how the progress is calculated. 

>  If we did not disable it I have to make poor hack in the code to work.

Not necessarily. We can simply refactor the method that updates the progress value. One idea: Have an object that, for each top-level element, tells us if it should be part of the progress counter or not. (This would also be super handy when we add pages that only include optional items in the future.)  
Is there anything speaking against this?

>  So I used disable() to remove it from control as suggested in https://angular.io/api/forms/AbstractControl#disable 

I think you have mis-understood the wordings used there. It says:
> Disables the control. This means the control is exempt from validation checks and excluded from the aggregate value of any parent. Its status is DISABLED.

This is really much more than just saying "I don't want this control not to be validated". It also says that this control is not part of the actual value. In general, you really only use this if you want to "grey out" a form element if the user is not supposed to enter data into this for now. (Of course, this is just one example of a pretty common use case.)

Reply:
Ya I agree disable() is more than avoiding validation. It will remove from the control itself which is the one I want to achieve. Also I feel the right way to do. Why id and isRated which are not filled by user to exist under control. For now it works without any hack. Adding flag at top level should work. But I don't think of a necessity now. We can refactor later on when we implement such pages.

Reply:
>For now it works without any hack. 

Using `getRawValue()` for this is a hack. It is just not the way you deal with form values. It is error-prone and may leak data: Other developers might think that their disabled entries will not be part of the final result. And this will then accidentally still be submitted to the server. By using getRawValue, you also simply disallow any developer from using the disable function in any part of the form in the way they are used to: It is simply not possible anymore to use "disable" on a form element in order to actually hide data from the final result.

But I got the feeling it makes no sense to argue here anymore. Just merge it like it is, but please at least add a prominent comment/info about the fact that disabled form controls are still submitted. This is not how forms are usually dealt with and any future developer on this code base may fall for this.

Btw: Another simple solution could have been to use a structure like this:

``` ts
interface MyData {
  id: number;
  isRated: boolean;
  submittedTime: Date;

  formValue: FormValue<FormsData>;
}
```

Then, when submitting/storing the data, you could simply have done the following:

```
const data: MyData = {
  id: theId,
  isRated: false,
  submittedTime: new Date(),

  formValue: this.formsData.value;
}
```

Reply:
A developer should be aware of what he is disabling. A need of disabling arrives when he doesn't want a field to be controlled. When he doesn't want to control? When it is not user entered fields. Also developer should refrain from adding unwanted fields in the interface other than some thing used for state maintainence in client  and system generated. These data structure are meant only for  local storage and should not be sent to the server directly. I don't think there will be a data leak if interface is properly handled. Let's say some fields has to enabled or disabled based on another formControl value. In this case if he does not want a field he can disable it. Here the default empty  value will be stored as done in the FormGroup instance creation for disabled fields. In this case if we use .value instead of getRawValues() the object will be no more of type in interface until we make it optional.

 Initially I thought of doing the same way as suggested in your interface, but it creates one more nested level of the object.

Reply:
> A need of disabling arrives when he doesn't want a field to be controlled. When he doesn't want to control? When it is not user entered fields.

If you don't want the user to enter anything, then just don't add it to the HTML.

There are use cases where you programatically fetch something and still want to have validators on that. Even if the actual input element is not directly seen on the UI.

> Also developer should refrain from adding unwanted fields in the interface other than some thing used for state maintainence in client and system generated.

What about the following use case? A checkbox asking "do you want to provide feedback to this (whatever)?". If it is checked, then some form items are displayed that the user now may fill out. But if the checkbox is not checked, then the user can just proceed. Do you think that, when the box is not checked, it really makes sense to include all of the disabled fields in the result?

From my point of view, this is a valid use case of using "disable()". 

Or, an example that is closer to our app: In our form we display a menu where the user may enter the schools and universities he graduated in. If an university is selected, then we ask for the name and, let's say, the grade of the final thesis. But if a school is selected, we only ask for the name. This is again one thing you would implement using the "disable()" function.

> Let's say some fields has to enabled or disabled based on another formControl value. In this case if he does not want a field he can disable it. Here the default empty value will be stored as done in the FormGroup instance creation for disabled fields.

The more elegant way is to simply not store the fields at all.


Finally, let's think about how this is handled in traditional HTML forms. If you wanted to store some data hidden from the user, then you would just create an input element with `type="hidden"`. In Angular, you simply don't create the HTML part of this form.

In traditional HTML forms, if you diabled an input field, it would not be sent to the server. In Angular, if you use "value", then the disabled inputs are not included. Coincidence?

The only reason why you are using these disabled fields in this PR is because of the progress indicator. If it didn't exist, then you wouldn't have had the need to disable the fields. So, from my point of view, one should simply fix the progress indicator instead.


But, I guess we should just keep going instead of wasting too much time on this discussion. You heard my points, I heard yours. I still would never ever use getRawValue(), but I cannot stop anybody from doing it.